lua/nvim-treesitter/ts_utils.lua

```diff
@@ -212,4 +214,117 @@ function M.get_node_at_cursor(winnr)
   return root:named_descendant_for_range(cursor[1]-1,cursor[2],cursor[1]-1,cursor[2])
 end
 
+-- Finds the definition node and it's scope node of a node
+-- @param node starting node
+-- @param bufnr buffer
+-- @returns the definition node and the definition nodes scope node
+function M.find_definition(node, bufnr)
+  local bufnr = bufnr or api.nvim_get_current_buf()  
+  local node_text = M.get_node_text(node)[1]
+  local current_scope = M.containing_scope(node)
+  local _, _, node_start = node:start()
+  local matching_def_nodes = {}
+
+  -- If a scope wasn't found then use the root node
+  if current_scope == node then
+    current_scope = parsers.get_parser(bufnr).tree:root()
+  end
+
+  -- Get all definitions that match the node text
+  for _, def in ipairs(locals.get_definitions(bufnr)) do
+    for _, def_node in ipairs(M.get_local_nodes(def)) do
+      if M.get_node_text(def_node)[1] == node_text then
+        table.insert(matching_def_nodes, def_node)       
+      end
+    end
+  end
+
+  -- Continue up each scope until we find the scope that contains the definition
+  while current_scope do
+    for _, def_node in ipairs(matching_def_nodes) do
+      if M.is_parent(current_scope, def_node) then
+        return def_node, current_scope
+      end
+    end
+    current_scope = M.containing_scope(current_scope:parent())
+  end
+
+  return nil, nil
```

Comment:
I would rather return `node, root` here, as it is actually the first occurence of this node, and we don't know where it belongs to.

Reply:
Sounds good